traitlets/config/loader.py

```diff
@@ -414,6 +423,21 @@ def _convert_to_config(self, dictionary):
         else:
             raise ValueError('Unknown version of JSON config file: {version}'.format(version=version))
 
+    def __enter__(self):
+        self.load_config()
+        return self.config
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """
+        Exit the context manager but do not handle any errors.
+
+        In case of any error, we do not want to write the potentially broken
+        configuration to disk.
+        """
+        with open(self.full_filename, 'w') as f:
+            f.write(json.dumps(self.config, indent=2))
```

Comment:
If there is an error here, the file will be left empty due to the `open(...'w')`. Perhaps dump the config prior to opening the file, so that it raises before deleting existing contents.